File: go/lmccstd/code.go
package lmccstd

// Raw-code spelling; no execution. See spec/vocab/format-code.md.
import (
	"fmt"
	"lmcc/lmcc"
	"strconv"
	"strings"
)

func codeIdentifier(s string) bool {
	if s == "" {
		return false
	}
	for i, c := range s {
		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_' || (i > 0 && c >= '0' && c <= '9')) {
			return false
		}
	}
	return true
}
func codeOptions(options *lmcc.Object, calls bool) (marker, tool string, err error) {
	marker, tool = "PY_END", "run_python"
	if options == nil {
		return
	}
	for _, k := range options.Keys {
		if k != "marker" && !(calls && k == "tool") {
			return "", "", fmt.Errorf("unknown option %q", k)
		}
	}
	if options.Has("marker") {
		marker, _ = options.Str("marker")
	}
	if options.Has("tool") {
		tool, _ = options.Str("tool")
	}
	if !codeIdentifier(marker) || !codeIdentifier(tool) {
		return "", "", fmt.Errorf("marker and tool must be nonempty ASCII identifiers")
	}
	return
}

type CodeArguments struct {
	base
	marker string
}

func newCodeArguments(options *lmcc.Object) (lmcc.Format, error) {
	marker, _, err := codeOptions(options, false)
	return &CodeArguments{base{"code_arguments"}, marker}, err
}
func (*CodeArguments) Accepts() []string           { return []string{"object"} }
func (*CodeArguments) Describe(*lmcc.Field) string { return "raw code" }
func codeValue(value any) (string, error) {
	obj, ok := value.(*lmcc.Object)
	if !ok || obj.Len() != 1 {
		return "", fmt.Errorf("code arguments must be exactly {code: string}")
	}
	code, ok := obj.Str("code")
	if !ok {
		return "", fmt.Errorf("code arguments must be exactly {code: string}")
	}
	return code, nil
}
func (c *CodeArguments) Write(value any, f *lmcc.Field) (any, error) {
	code, err := codeValue(value)
	if err != nil {
		return nil, err
	}
	if strings.Contains(code, c.marker) {
		return nil, &lmcc.Error{Code: "value-collides", Detail: "code contains heredoc marker " + c.marker + "; choose another marker"}
	}
	return code, nil
}
func (c *CodeArguments) Read(span lmcc.Span, f *lmcc.Field) (any, error) {
	var body strings.Builder
	for _, raw := range span.Parts {
		p, ok := raw.(*lmcc.Object)
		if !ok {
			return nil, fmt.Errorf("code arguments need text parts")
		}
		typ, _ := p.Str("type")
		text, ok := p.Str("text")
		if typ != "text" || !ok {
			return nil, fmt.Errorf("code arguments need text parts")
		}
		body.WriteString(text)
	}
	code := body.String()
	if strings.Contains(code, c.marker) {
		return nil, fmt.Errorf("code contains heredoc marker %q", c.marker)
	}
	return lmcc.Obj("code", code), nil
}

type CodeCalls struct {
	base
	tool      string
	arguments *CodeArguments
}

func newCodeCalls(options *lmcc.Object) (lmcc.Format, error) {
	marker, tool, err := codeOptions(options, true)
	return &CodeCalls{base{"code_calls"}, tool, &CodeArguments{base{"code_arguments"}, marker}}, err
}
func (*CodeCalls) Accepts() []string           { return []string{"list[*]", "*"} }
func (*CodeCalls) Emits() string               { return "parts" }
func (*CodeCalls) Reads() []string             { return []string{"text", "tool_call"} }
func (*CodeCalls) Describe(*lmcc.Field) string { return "heredoc tool calls" }
func (c *CodeCalls) native(call any, f *lmcc.Field) (*lmcc.Object, error) {
	obj, ok := call.(*lmcc.Object)
	if !ok {
		return nil, fmt.Errorf("expected a call object")
	}
	name, _ := obj.Str("name")
	id, ok := obj.Str("id")
	if name != c.tool || !ok || id == "" {
		return nil, fmt.Errorf("expected a %q call with a nonempty id", c.tool)
	}
	code, err := codeValue(obj.Object("input"))
	if err != nil {
		return nil, err
	}
	args, err := c.arguments.Read(lmcc.SpanOfText(code), f)
	if err != nil {
		return nil, err
	}
	return lmcc.Obj("id", id, "name", name, "input", args), nil
}
func (c *CodeCalls) Write(value any, f *lmcc.Field) (any, error) {
	list, ok := value.([]any)
	if !ok {
		return nil, fmt.Errorf("calls must be a list")
	}
	out := []any{}
	for _, call := range list {
		v, err := c.native(call, f)
		if err != nil {
			return nil, err
		}
		p := lmcc.Obj("type", "tool_call")
		for _, k := range v.Keys {
			x, _ := v.Get(k)
			p.Set(k, x)
		}
		out = append(out, p)
	}
	return out, nil
}
func (c *CodeCalls) Read(span lmcc.Span, f *lmcc.Field) (any, error) {
	calls := []any{}
	for _, raw := range span.Parts {
		p, ok := raw.(*lmcc.Object)
		if !ok {
			return nil, fmt.Errorf("expected a part object")
		}
		typ, _ := p.Str("type")
		if typ == "tool_call" {
			v, err := c.native(p, f)
			if err != nil {
				return nil, err
			}
			calls = append(calls, v)
		} else {
			args, err := c.arguments.Read(lmcc.Span{Parts: []any{p}}, f)
			if err != nil {
				return nil, err
			}
			calls = append(calls, lmcc.Obj("id", "call_"+strconv.Itoa(len(calls)+1), "name", c.tool, "input", args))
		}
	}
	return calls, nil
}

func heredocTools(options *lmcc.Object) (*lmcc.Strategy, error) {
	marker, tool, err := codeOptions(options, true)
	if err != nil {
		return nil, err
	}
	opening, closing := tool+" <<'"+marker+"'\n", "\n"+marker
	s := lmcc.NewStrategy()
	s.Requires = []string{"instruct"}
	s.Visible = false
	s.Placement = lmcc.Obj("@role", "message:system")
	s.Via = lmcc.Obj("@role", "tool_catalog")
	s.Fragments = lmcc.Obj("system", "To request "+tool+", emit this heredoc and wait for its result:\n"+opening+"<code>"+closing+"\nDo not put "+marker+" anywhere in the code. Otherwise reply normally.")
	s.Routings = []*lmcc.Object{lmcc.Obj("from", "text", "between", []any{opening, closing}, "to", "@role.calls", "consume", true, "suffices", true)}
	s.Turns = lmcc.Obj("call", "{name} <<'"+marker+"'\n{input}"+closing,
		"result", "Result of {name} ({id}):\n{output}",
		"input_format", lmcc.Obj("use", "code_arguments", "options", lmcc.Obj("marker", marker)),
		"probe", lmcc.Obj("name", tool, "input", lmcc.Obj("code", "print(6 * 7)\n")))
	return s, nil
}

File: python/lmcc_std/code.py
"""Raw-code argument spelling and heredoc calls (spec/vocab/format-code.md).

No execution. Arguments and capture bodies stay exact, including indentation
and trailing newlines. The envelope is owned by the strategy.
"""

import dataclasses

from lmcc import core
from lmcc.errors import refuse
from lmcc.formats import Format
from lmcc.strategy import Strategy

from .formats import lift

VERSION = "0.1.0"


def _options(options, *, calls=False):
    allowed = {"marker", "tool"} if calls else {"marker"}
    if set(options) - allowed:
        raise ValueError(f"unknown options: {sorted(set(options) - allowed)}")
    marker, tool = options.get("marker", "PY_END"), options.get("tool", "run_python")
    for key, value in (("marker", marker), ("tool", tool)):
        if (not isinstance(value, str) or not value or not value.isascii()
                or value[0] not in "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
                or any(c not in "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789" for c in value)):
            raise ValueError(f"{key} must be a nonempty ASCII identifier")
    return marker, tool


class CodeArguments(Format):
    accepts = ("object",)

    def __init__(self, options):
        self.marker, _ = _options(options)

    def describe(self, field):
        return "raw code"

    def _code(self, value):
        if not isinstance(value, dict) or set(value) != {"code"} or not isinstance(value["code"], str):
            raise ValueError("code arguments must be exactly {code: string}")
        return value["code"]

    def write(self, value, field):
        code = self._code(value)
        if self.marker in code:
            refuse("value-collides", f"code contains heredoc marker {self.marker!r}; choose another marker")
        return code

    def read(self, span, field):
        if any(p.get("type") != "text" or not isinstance(p.get("text"), str) for p in span.parts):
            raise ValueError("code arguments need text parts")
        code = "".join(p["text"] for p in span.parts)  # NOT span.text: code whitespace is data
        if self.marker in code:
            raise ValueError(f"code contains heredoc marker {self.marker!r}")
        return {"code": code}


class CodeCalls(Format):
    accepts = ("list[*]", "*")
    emits = "parts"
    reads = ("text", "tool_call")

    def __init__(self, options):
        marker, self.tool = _options(options, calls=True)
        self.arguments = CodeArguments({"marker": marker})

    def describe(self, field):
        return "heredoc tool calls"

    def _native(self, call, field):
        call = dataclasses.asdict(call) if dataclasses.is_dataclass(call) else call
        if (not isinstance(call, dict) or call.get("name") != self.tool
                or not isinstance(call.get("id"), str) or not call["id"]):
            raise ValueError(f"expected a {self.tool!r} call with a nonempty id")
        body = self.arguments._code(call.get("input"))
        # Read-side validation raises a format error, not a render refusal.
        self.arguments.read(core.Span.of_text(body), field)
        return {"id": call["id"], "name": self.tool, "input": {"code": body}}

    def write(self, value, field):
        if not isinstance(value, list):
            raise ValueError("calls must be a list")
        return [{"type": "tool_call", **self._native(c, field)} for c in value]

    def read(self, span, field):
        calls = []
        for part in span.parts:
            if part.get("type") == "tool_call":
                calls.append(self._native(part, field))
            else:
                args = self.arguments.read(core.Span([part]), field)
                calls.append({"id": f"call_{len(calls) + 1}", "name": self.tool, "input": args})
        return lift(field.annotation, calls)


def heredoc_tools(options):
    marker, tool = _options(options, calls=True)
    opening, closing = f"{tool} <<'{marker}'\n", f"\n{marker}"
    return Strategy(
        requires=["instruct"], visible=False,
        placement={"@role": "message:system"}, via={"@role": "tool_catalog"},
        fragments={"system": f"To request {tool}, emit this heredoc and wait for its result:\n"
                   f"{opening}<code>{closing}\n"
                   f"Do not put {marker} anywhere in the code. Otherwise reply normally."},
        routings=[{"from": "text", "between": [opening, closing], "to": "@role.calls",
                   "consume": True, "suffices": True}],
        turns={"call": "{name} <<'" + marker + "'\n{input}" + closing,
               "result": "Result of {name} ({id}):\n{output}",
               "input_format": {"use": "code_arguments", "options": {"marker": marker}},
               "probe": {"name": tool, "input": {"code": "print(6 * 7)\n"}}})


def install(registry, *, exist_ok=True):
    registry.register_format("code_arguments", CodeArguments, version=VERSION, exist_ok=exist_ok)
    registry.register_format("code_calls", CodeCalls, version=VERSION, exist_ok=exist_ok)
    registry.register_strategy("heredoc_tools", heredoc_tools, version=VERSION, exist_ok=exist_ok)
